core/src/main/java/nl/nn/adapterframework/core/PeekableDataIterator.java

```diff
@@ -0,0 +1,56 @@
+/*
+   Copyright 2021 WeAreFrank!
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.adapterframework.core;
+
+public class PeekableDataIterator<I> implements IDataIterator<I> {
+
+	private IDataIterator<I> target;
+	private I peeked = null;
+	
+	public PeekableDataIterator(IDataIterator<I> target) {
+		super();
+		this.target = target;
+	}
+
+	@Override
+	public boolean hasNext() throws SenderException {
+		return peeked!=null || target.hasNext();
+	}
+
+	@Override
+	public I next() throws SenderException {
+		
+		if (peeked!=null) {
+			I result = peeked;
+			peeked = null;
+			return result;
+		}
+		return target.next();
+	}
+
+	public I peek() throws SenderException {
+		if (peeked==null && hasNext()) {
+			peeked=target.next();
+		}
+		return peeked;
+	}
```

Comment:
Een beetje javadoc is handig denk ik:
Returns peeked object if present, or NULL if not. Can be peeked multiple times.